test/fuzz/json/json.go

```diff
@@ -0,0 +1,143 @@
+	nonstrictOpt := SerializerOptions{false, false, true}
+	scheme := runtime.NewScheme()
+	strictSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, strictOpt)
+	nonstrictSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, nonstrictOpt)
+	ysSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, strictYamlOpt)
+	psSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, strictPrettyOpt)
+	obj0, _, err0 := strictSer.Decode(data, gvk, nil)
+	obj1, _, err1 := nonstrictSer.Decode(data, gvk, nil)
+	obj2, _, err2 := ysSer.Decode(data, gvk, nil)
+	obj3, _, err3 := psSer.Decode(data, gvk, nil)
+	if obj0 == nil {
+		if obj1 != nil {
+			panic("NonStrict is stricter than Strict")
+		}
+		if obj2 != nil {
+			panic("Yaml strict different from plain strict")
+		}
+		if obj3 != nil {
+			panic("Pretty strict different from plain strict")
+		}
+		if err0 == nil || err1 == nil || err2 == nil || err3 == nil {
+			panic("no error")
+		}
+		return 0
+	}
+
+	if err0 != nil {
+		panic("got object and error for strict")
+	}
+	if err2 != nil {
+		panic("got object and error for yaml strict")
+	}
+	if err3 != nil {
+		panic("got object and error pretty strict")
+	}
+
+	var b0 bytes.Buffer
+	err4 := strictSer.Encode(obj0, &b0)
+	if err4 != nil {
+		panic("Can't encode decoded data")
+	}
+	if !bytes.Equal(b0.Bytes(), data) {
+		panic("Encoded data doesn't match original")
+	}
+
+	b0.Reset()
+	err5 := ysSer.Encode(obj1, &b0)
+	if err5 != nil {
+		panic("Can't encode yaml strict decoded data")
+	}
+	if !bytes.Equal(b0.Bytes(), data) {
+		panic("Encoded yaml strict data doesn't match original")
+	}
+
+	b0.Reset()
+	err6 := psSer.Encode(obj2, &b0)
+	if err6 != nil {
+		panic("Can't encode pretty strict decoded data")
+	}
+	if !bytes.Equal(b0.Bytes(), data) {
+		panic("Encoded pretty strict data doesn't match original")
+	}
+
+	b0.Reset()
+	err7 := nonstrictSer.Encode(obj3, &b0)
+	if err7 != nil {
+		panic("Can't encode nonstrict decoded data")
+	}
+	if !bytes.Equal(b0.Bytes(), data) {
+		panic("Encoded nonstrict data doesn't match original")
+	}
+	return 1
+}
+
+func FuzzNonStrictDecode(data []byte) int {
+	gvk := &schema.GroupVersionKind{Version: "v1"}
+	nonstrictOpt := SerializerOptions{false, false, true}
+	nonstrictYamlOpt := SerializerOptions{true, false, true}
+	nonstrictPrettyOpt := SerializerOptions{false, true, true}
+	scheme := runtime.NewScheme()
+	nonstrictSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, nonstrictOpt)
+	ynsSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, nonstrictYamlOpt)
+	pnsSer := NewSerializerWithOptions(DefaultMetaFactory, scheme, scheme, nonstrictPrettyOpt)
+	obj0, _, err0 := nonstrictSer.Decode(data, gvk, nil)
+	if err0 != nil {
+		return 0
+	}
+
+	var b0 bytes.Buffer
+	err1 := nonstrictSer.Encode(obj0, &b0)
+	if err1 != nil {
+		panic("Can't encode decoded data")
+	}
+	if !bytes.Equal(b0.Bytes(), data) {
```

Comment:
I wouldn't expect unknown fields to be preserved when using nonstrict.

Reply:
Fair point. Removed but added some additional encode steps.